### ISSUE TITLE
`rendercomplete` called multiple times with incomplete maps rendered

### ISSUE THREAD

**Issue by perliedman** [openlayers/openlayers#15236]
**Describe the bug**

I have some code intended to export an OpenLayers map to a static image, for use in a PDF/print export.

I have followed the basic outlines of the [Map Export](https://openlayers.org/en/latest/examples/export-map.html) example (the actual use case is quite a bit more complex, but the basic steps are the same).

Doing this, I have encountered a weird problem when trying to print layers based on `RasterSource`, where I frequently end up with images that miss some/most/all tiles. 

**To Reproduce**

1. Go to https://codesandbox.io/s/competent-moore-xj67kc?file=/src/index.mjs
2. A world map is displayed as a static image; this image is produced by OpenLayers from OSM data fed through a simple `RasterLayer` operation that filters out the sea (sort of)
3. Notice that the `rendercomplete` message is logged to the console 10 times (for me, might vary depending on machine I guess)
4. If the code is changed to use `map.once('rendercomplete', ...)`, like in the original Map Export example, a completely white image is instead generated
5. Looking at the results of each `rendercomplete`, it looks like tiles are faded in from completely white to fully opaque
6. If the `size` is changed to a signficantly larger value (like 3200x3200 for high resolution print), the number of `renderComplete` is increased dramatically, and the time between them is so long that the debounce trick I've used doesn't work, I get "intermediate" rendering results with some (often majority) of tiles missing

**Expected behavior**

I expect a usable map image on the first `rendercomplete`, so that `map.once` can be used instead of debounce tricks, or at least a way to figure out if a render result is actually complete, or if I can expect more data to be rendered.

In my real world application, I've seen cases where the render loop seemingly _never_ stabilizes, producing infinite number of `rendercomplete`s.

I confess I'm not sure if I can realistically expect this to work, if I misunderstood some feature when writing this code, or if there's some inherent technical limitation in creating static hi-res images from OpenLayers maps with intricate `RasterSource`s, so any help on this would be very helpful!

**Comment by mike-000** [comment]
It looks like any of the OpenLayers examples which use RasterSource would have the same problem if used in an export.  There might be better alternative to a RasterSource,  e.g. using the layer's `postrender` event (as in https://openlayers.org/en/latest/examples/image-filter.html)

```
const layer = new TileLayer({
  source: new OSM()
});

layer.on('postrender', function (event) {
  const context = event.context
  const canvas = context.canvas;
  const width = canvas.width;
  const height = canvas.height;
  const imageData = context.getImageData(0, 0, width, height);
  const data = imageData.data;
  for (let i = 0, ii = data.length; i < ii; i += 4) {
    if (data[i] < 224) {
      data[i + 3] = 0;
    }
  }
  context.putImageData(imageData, 0, 0);
});
```

or using a WebGL layer and style

```
const layer = new WebGLTileLayer({
  source: new OSM(),
  style: {
    color: [
      'array',
      ['band', 1],
      ['band', 2],
      ['band', 3],
      ['case', ['<', ['band', 1], 224 / 255], 0, ['band', 4]]
    ]
  }
});
```

**Comment by ahocevar** [comment]
So it looks like there is a bug with Raster sources - they prevent the `rendercomplete` event from working correctly. But note that this could also be due to browser limitations. A destination image size of 3200x3200 pixels means that the Raster source needs to create an additional canvas with the same size, and that size might be 6400x6400 pixels or even more on a Retina display, which is likely to exceed the maximum canvas size allowed by the browser.

**Comment by perliedman** [comment]
@mike-000 thanks for the input! I might look into the possibility of using WebGL which might be better from both performance and maybe these days it's a more well tested path. Having said that, my real application does a lot more complex operations than just filtering on a single band, so it might take significant work to rewrite the current setup.

**Comment by perliedman** [comment]
@ahocevar agreed that 3200x3200 might be very large, and in my real application I don't think we use more than 1600x1600, but still see this issue. The same setup does work without issues (but probably lots of `rendercomplete` events) when we run it as an "interactive" map on screen, even if I use the whole screen (more pixels than 1600x1600 but less than 3200x3200).

A bit curious about the OOM scenario: wouldn't I see some error for this case? I imagine getting the canvas context would fail if the browser can't allocate the memory? Or at least some error/warning logged to the console. I guess what I'm saying is that if rendering an image like this is unrealistic, it would be great to get some indication that what I'm requesting is impossible.
